src/v/archival/ntp_archiver_service.cc

```diff
@@ -645,6 +643,50 @@ ntp_archiver::download_manifest() {
     co_return std::make_pair(tmp, result);
 }
 
+ss::future<> ntp_archiver::maybe_upload_manifest() {
+    if (
+      _parent.archival_meta_stm()->get_dirty()
+      == cluster::archival_metadata_stm::state_dirty::clean) {
+        vlog(_rtclog.trace, "Manifest is clean, skipping upload");
+        co_return;
+    }
+
+    auto upload_insync_offset
+      = _parent.archival_meta_stm()->get_insync_offset();
+    vlog(
+      _rtclog.trace,
+      "Uploading partition manifest, insync_offset={}",
+      upload_insync_offset);
+
+    auto result = co_await upload_manifest();
+    if (result == cloud_storage::upload_result::success) {
+        _last_manifest_upload_time = ss::lowres_clock::now();
+        auto deadline = ss::lowres_clock::now()
+                        + config::shard_local_cfg()
+                            .cloud_storage_metadata_sync_timeout_ms.value();
+        auto errc = co_await _parent.archival_meta_stm()->mark_clean(
```

Comment:
This will double the amount of archival metadata batches that we have to replicate. Maybe it is possible to batch this `mark_clean` batch with `add_segments` batches from the next iteration or something like that? The `archival_metadata_stm` supports batching so all this commands can be placed into a single batch and replicated together.

Reply:
So basically, to upload something we will have to 
1) upload
2) replicate first batch
3) upload manifest
4) replicate second batch

all steps are sequential.

Reply:
If we upload the manifest optimistically while segments uploads are ongoing, how do we deal with segment upload failures? We can detect this locally, and keep the manifest dirty, but read replicas will still see a manifest with phantom segments for a while.

Reply:
The uploaded manifest only ever contains persisted data from the stm: that means that when we do it in parallel with segment uploads, it doesn't include the segments' we're currently uploading.  That's fine at high throughput, because we will soon be uploading more segments and the manifest will get updated then.  At lower throughputs there is a different case where after uploading some segments, if the stm transitioned clean->dirty, then we will upload a manifest _after_ uploading the segments, so that it includes the just-uploaded segments promptly.

Reply:
> The uploaded manifest only ever contains persisted data from the stm: that means that when we do it in parallel with segment uploads, it doesn't include the segments' we're currently uploading.

Then it makes sense. Must have misunderstood the intent from your comment.